mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp

```diff
@@ -695,6 +695,10 @@ void SYCLHostScheduleKernel::getEffects(
     SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
         &effects) {
   auto *defaultResource = SideEffects::DefaultResource::get();
+  // Unconditionally add a write effect on the handler to prevent the op from
```

Comment:
I am ok with the current solution, but thinking if adding `ConditionallySpeculatable` would solve the problem.

Reply:
It would still be trivially dead, so not sure that would fix it

Reply:
shouldn't removing a speculatable operation be change of behavior, and cannot be safely removed?

Reply:
Sounds like MemoryEffectOpInterface should be improved to indicate it has memory effect outside of its arguments.

Reply:
Approve in the meantime.

Reply:
> Sounds like MemoryEffectOpInterface should be improved to indicate it has memory effect outside of its arguments. WDYT?

They recommend using an abstract resouce in that case ([see](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/#modeling)), but that would break our reaching definition analysis.